File: calculator/rubycalc/lib/parser.rb
require 'ast'
require 'scanner'
require 'token'
require 'calcex'

class Parser
  def initialize(istream)
    @scan = Scanner.new(istream)
  end
   
  def parse()
    return Prog()
  end
  
  #private public protected
  def Prog()
    result = Expr()
    t = @scan.getToken()
    
    unless t.type == :eof then
      print "Expected EOF. Found ", t.type, ".\n"
      raise ParseError.new
    end
    
    return result
  end
  
  def Expr() 
    RestExpr(Term())
  end
   
  def RestExpr(e) 
    t = @scan.getToken()
    
    if t.type == :add then
         return RestExpr(AddNode.new(e,Term()))
    end
    
    if t.type == :sub then
      return RestExpr(SubNode.new(e,Term()))
    end
      
    @scan.putBackToken()
    
    e
  end
  
  def Term()
    RestTerm(Storable())
    # Write your Term() code here. This code is just temporary
    # so you can try the calculator out before finishing it.
    
    # t = @scan.getToken()
    
    # if t.type == :number then
    #   val = t.lex.to_i
    #   return NumNode.new(val)
    # end
    
    # puts "Term not implemented\n"
    
    # raise ParseError.new
  end
   
  def RestTerm(e)
    t = @scan.getToken()

    if t.type == :times then
      return RestTerm(TimesNode.new(e, Storable()))
    end
    if t.type == :divide then
      return RestTerm(DivideNode.new(e, Storable()))
    end
    @scan.putBackToken
    
    e
  end
   
  def Storable()
    fact = Factor()
    t = @scan.getToken

    if t.type == :keyword then
      if t.lex == "S" then
        return StoreNode.new(fact)
      else
        raise ParseError.new
      end
    end

    @scan.putBackToken

    fact
  end
   
  def Factor() 
    t = @scan.getToken
    
    if t.type == :number then
      return NumNode.new(t.lex.to_i)
    end
    if t.type == :keyword then
      if t.lex == "R" then
        return RecallNode.new
      else
        raise ParseError.new
      end
    end
    if t.type == :lparen then
      expr = Expr()
      t = @scan.getToken

      if t.type == :rparen then
        return expr
      else
        raise ParseError.new
      end
    end
    
    raise ParseError.new
  end         
end

File: calculator/cppcalc/parser.cpp
#include "parser.h"
#include "calcex.h"
#include <string>
#include <sstream>

Parser::Parser(istream* in) {
   scan = new Scanner(in);
}

Parser::~Parser() {
   try {
      delete scan;
   } catch (...) {}
}

AST* Parser::parse() {
   return Lineas();
}

AST* Parser::Lineas() {
   AST* result = Linea();
   Token* t = scan->getToken();

   return Lineas();
}

AST* Parser::Linea() {
  Token* t = scan->getToken();
  if(t->getType() == keyword){
    if(t->getLex() == "let"){
      
    }
  }
  if(t->getType() == eof){
    //Code here
  }
  return RestExpr(Term());
}

AST* Parser::RestExpr(AST* e) {
   Token* t = scan->getToken();

   if (t->getType() == add) {
      return RestExpr(new AddNode(e,Term()));
   }

   if (t->getType() == sub){
      return RestExpr(new SubNode(e,Term()));
   }
   scan->putBackToken();

   return e;
}

AST* Parser::Term() {
  return RestTerm(Storable());
}

AST* Parser::RestTerm(AST* e) {
  Token* t = scan->getToken();

  if(t->getType() == times){
    return RestTerm(new TimesNode(e, Storable()));  
  }

  if(t->getType() == divide){
    return RestTerm(new DivideNode(e, Storable()));
  }

  if(t->getType() == mod){
    return RestTerm(new ModNode(e, Storable()));
  }

  scan->putBackToken();

  return e;
}

AST* Parser::Storable() {
  return MemOperation(Negative());
}

AST* Parser::MemOperation(AST* e){
  Token* t = scan->getToken();
  if(t->getType() == keyword){
    if(t->getLex() == "S"){
      return new StoreNode(e);
    }
    if(t->getLex() == "R"){
      return new RecallNode();
    }
    if(t->getLex() == "M"){
      return new MemoryNode();
    }
  }
  scan->putBackToken();
  return e;
}

AST* Parser::Negative(){
  AST* res = Factor();
  Token* t = scan->getToken();
  if(t->getType == negative){
    if(t->getLex() == "-"){
      //Code here
    }
  }
  scan->putBackToken();
  return res;
}

AST* Parser::Factor() {
  Token* t = scan->getToken();
  if(t->getType() == number){
    istringstream in(t->getLex());
    int val;
    in >> val;
    return new NumNode(val);
    // return new NumNode(stoi(t->getLex())); en c++ 11
  }
  if(t->getType() == keyword){
    if(t->getLex() == "R"){
      return new RecallNode();
    }
  }
  if(t->getType() == keyword){
    if(t->getLex() == "C"){
      return new ClearNode();
    }
  }
  if(t->getType() == indentifier){
    
  }
  if(t->getType() == lparen){
    //No retorno Expr() de una vez, porque la gramatica pide un token mas, rparen
    AST* res = Expr();
    t = scan->getToken();
    if(t->getType() == rparen){
      return res;
    }
    cerr << "Eror, expected: )" << endl;
    throw ParseError;
  }
  cerr << "Error, expected number | R | (" << endl;
  throw ParseError;
}
   
